src/WebJobs.Script.WebHost/Security/SecretManager.cs

```diff
@@ -83,24 +83,24 @@ public virtual HostSecretsInfo GetHostSecrets()
                 }
                 else
                 {
-                    // initialize with new secrets and save it
-                    hostSecrets = new HostSecrets
-                    {
-                        MasterKey = GenerateSecret(DefaultMasterKeyName),
-                        FunctionKeys = new List<Key>
-                        {
-                            GenerateSecret(DefaultFunctionKeyName)
-                        }
-                    };
+                    hostSecrets = GenerateHostSecrets(secretFilePath);
+                }
 
-                    string secretContent = ScriptSecretSerializer.SerializeHostSecrets(hostSecrets);
-                    File.WriteAllText(secretFilePath, secretContent);
+                // Host secrets will be in the original persisted state at this point (e.g. encrypted),
+                // so we read the secrets running them through the appropriate readers
+                hostSecrets = ReadHostSecrets(hostSecrets);
+
+                // If the persistence state of any of our secrets is stale (e.g. the encryption key has been rotated), update
+                // the state and persist the secrets
+                if (hostSecrets.MasterKey.PersistenceStateStale || hostSecrets.FunctionKeys.Any(k => k.PersistenceStateStale))
```

Comment:
Perhaps move this logic into HostSecrets, allowing you to do hostSecrets.RequiresMigration (or whatever name you decide)